### DISCUSSION TITLE
Get dependency via DI in wrapper function

### DISCUSSION THREAD

**Discussion by Soerboe** [fastapi/fastapi#6285] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from functools import wraps

import uvicorn as uvicorn
from fastapi import FastAPI, Depends

app = FastAPI()


def add_id_header(func):
    @wraps(func)
    async def wrapper(*args, **kwargs):
        name_id = func(*args, **kwargs)
        headers = {"etag": name_id}

        # How can I get access to the Response here?
        # response.headers.update(headers)

    return wrapper


def generate_id(name: str):
    return name.lower().replace(' ', '')


@app.get('/hello/{name}', dependencies=[Depends(add_id_header(generate_id))])
def get(name: str):
    return f'Hello, {name}'
```

### Description

I'd like to create a structure where I can create a function (in example code `generate_id`) which takes arguments via dependency injection. In addition I would like to run arbitrary code after that function which also should get dependency injected arguments in some way. In the example code this is done by using a wrapper function. However, I can not get access to the response in the wrapper function (which should be injected via DI). Is there a way to do this with the current setup (using wrapper function) or is there some other way to accomplish this?

If using class based dependencies it would have worked if the arguments to `__init__` could also be injected via DI.

### Operating System

Linux

### FastAPI Version

0.73.0

### Python Version

3.10.6

**Comment by JarroVGIT** [comment]
Hmm, I don't think `wraps` is the right route here. If I understand you correctly, you are looking for a way to alter the response before it is returned to the client (e.g. the enduser) and run some other code, correct? 

I can think of two ways to deal with this. First, you could use [middleware](https://fastapi.tiangolo.com/tutorial/middleware/). You will receive the request prior to FastAPI receiving it, and will get the response that FastAPI returns before your middleware will pass it to the calling object (uvicorn or some other middleware). You can mutate both to your liking. 

However, if you want your code to be executed on certain paths (or groups of paths), I would recommend using dependencies. Dependencies can also be used to alter responses, **and to run code after the response is sent!** This latter functionality is accomplished by using `yield` rather than `return` in your dependency. See the below example on how to set headers on a response using a dependency, and how to execute code after returning the response. The code can be ran as-is.

```python
from fastapi import Depends, FastAPI, Request, Response
from functools import wraps

app = FastAPI()


def dependency(response: Response):
    print("1. Dependency is called first..")
    response.headers["id"] = "hello"
    print("2. Added a header to the response object, now yield from dependency..")
    yield
    print(
        "4. Any code after yield in the dependency is ran after the response has returned. "
    )


@app.get("/", dependencies=[Depends(dependency)])
async def root():
    print("3. This is executed in your endpoint, after all dependencies have yielded..")
    return {"hello": "world"}


if __name__ == "__main__":
    import uvicorn

    uvicorn.run(app, host="0.0.0.0", port=8000)
```
The logs will show this if you visit the `localhost:8000` endpoint:


```log
1. Dependency is called first..
2. Added a header to the response object, now yield from dependency..
3. This is executed in your endpoint, after all dependencies have yielded..
INFO:     127.0.0.1:56410 - "GET / HTTP/1.1" 200 OK
4. Any code after yield in the dependency is ran after the response has returned. 
```
You don't have to use `wraps` at all in this setup, which would greatly simplify your codebase.

**Reply by YuriiMotov** [reply]
In current version it's a bit different:
```
1. Dependency is called first..
2. Added a header to the response object, now yield from dependency..
3. This is executed in your endpoint, after all dependencies have yielded..
4. Any code after yield in the dependency is ran after the response has returned. 
INFO:     127.0.0.1:37560 - "GET / HTTP/1.1" 200 OK
```

So, the code after `yield` is executed after exiting endpoint function but before sending the response

**Comment by Soerboe** [comment]
@JarroVGIT Thanks for the reply. I should maybe have been more clear of what I am aiming for. BTW, I know how dependencies and middleware works but I have not been able to find a solution still.

The reason for using a wrapper in the first place is because the wrapper function (`add_id_header`) will be a general one coming from a core component or a plugin. The wrapped function (`generate_id`) will be specific to each group of endpoints (routers). I need to get the return value from the `generate_id` function in the `add_id_header` function. And then both functions must be able to leverage the dependency injection system.

One way to work around the issue is to add all the needed dependencies for `add_id_header` function also in the `generate_id`  function, but this is sub-optimal as they will mostly be unused in the `generate_id` function. Like this:

```python
def add_id_header(func):
    @wraps(func)
    async def wrapper(response: Response, *args, **kwargs):
        name_id = func(response, *args, **kwargs)
        headers = {"etag": name_id}

        response.headers.update(headers)

    return wrapper


def generate_id(response: Response, name: str): # response unused
    return name.lower().replace(' ', '')
``` 

I have tried to think about some other way to handle this but I have not found one. Any input would be valuable!

**Comment by JarroVGIT** [comment]
I am not sure if I completely understand your use case, so my apologies if this is not helpful. You could create a common dependency class that is initiated with a function parameter, and then the function is called when the common dependency class is called. Again, I am sorry if this is not what you are looking for. In general: if you want the dependency system to work, then FastAPI must be able to 'follow' dependency parameters with either `Depends()` as default parameter values, or specific paramater types (such as `Response` and `Request`). 

```python
    from typing import Any, Callable

    from fastapi import APIRouter, Depends, FastAPI, Response

    app = FastAPI()


    class MyCommonDependencyClass:
        def __init__(self, dependency_func: Callable[..., Any] = None) -> None:
            self.dependency_func = dependency_func

        def __call__(self, response: Response) -> Any:
            print("this is from the common dependency class")
            self.dependency_func(response)


    def dependency_function_for_router_1(response: Response):
        print("this is from dependency_function_for_router_1")
        response.headers["hello"] = "world1"


    def dependency_function_for_router_2(response: Response):
        print("this is from dependency_function_for_router_2")
        response.headers["hello"] = "world2"


    @app.get("/")
    async def root():
        return {"hello": "world"}


    dependency_for_router_1 = MyCommonDependencyClass(dependency_function_for_router_1)
    router1 = APIRouter(prefix="/router1")


    @router1.get("/", dependencies=[Depends(dependency_for_router_1)])
    async def from_router1():
        return "hello world1"


    dependency_for_router_2 = MyCommonDependencyClass(dependency_function_for_router_2)
    router2 = APIRouter(prefix="/router2")


    @router2.get("/", dependencies=[Depends(dependency_for_router_2)])
    async def from_router2():
        return "hello world2"


    app.include_router(router1)
    app.include_router(router2)

    if __name__ == "__main__":
        import uvicorn

        uvicorn.run(app, host="0.0.0.0", port=8000)
```

**Comment by Soerboe** [comment]
@JarroVGIT This is what I am aiming for (and also sort of what I'm doing but with wrapper function instead of callable class). I have just one issue though.

I would like to be able to have different dependency injected arguments for `dependency_function_for_router_1` and `dependency_function_for_router_2` (and not Response as I don't need it in these functions). E.g.

```python
def dependency_function_for_router_1(path_param_1, crud_router1):
        return crud_router1.get(path_param_1)


def dependency_function_for_router_2(path_param_2, crud_router2):
        return crud_router2.get(path_param_2)
```

Is there a way to do that?

**Comment by JarroVGIT** [comment]
Hmm I see, I don't think that is possible as you would have to declare the parameters in all (nested) callables for them to be available in there.

**Comment by Soerboe** [comment]
Alright. Then I guess my best option is to use the workaround in https://github.com/tiangolo/fastapi/issues/5279#issuecomment-1219134442.

**Comment by Soerboe** [comment]
@JarroVGIT I found a way easier solution with the help of https://github.com/zhanymkanov/fastapi-best-practices#4-chain-dependencies

```python
import uvicorn as uvicorn
from fastapi import FastAPI, Depends
from starlette.responses import Response


app = FastAPI()


def add_id_header(func):

    async def wrapper(response: Response, name_id=Depends(func)):
        headers = {"etag": name_id}
        response.headers.update(headers)

    return wrapper


def generate_id(name: str):
    return name.lower().replace(' ', '')


@app.get('/hello/{name}', dependencies=[Depends(add_id_header(generate_id))])
def get(name: str):
    return f'Hello, {name}'


if __name__ == '__main__':
    uvicorn.run("main:app", host="127.0.0.1", port=8008, reload=True, debug=True)

```

**Comment by JarroVGIT** [comment]
That's awesome! I really never would've thought that would work, because I thought the return of `inspect.signature()` on `add_id_header` would never retain the additional parameters of the wrapper that is returned. 

But when I think about it now it actually does make sense. We normally pass `function_name` without parenthesis, so as an object, to `Depends()`, but passing a `outer_function(inner_function_object)` would actually _execute_ `outer_function` with param `inner_functino_object`, returning a function object. That last function object, has the `Response` and another `Depends()` param which is then (correctly) picked up by the FastAPI resolver of dependencies.

That's really cool stuff, thanks for sharing! If you don't mind, I might actually write a blog on that.

EDIT: if anyone is interested in how this works, consider the below example (dependency resolving in FastAPI is build upon the inspection of the parameters in the path operation functions and their default values (if Depends(), then resolve that as well etc). It illustrate why this works in FastAPI:

```python
import inspect
from typing import Any, Callable


def outer_function(inner_function: Callable[..., Any]):
    def wrapper(a: int, b: int):
        return inner_function(a, b)

    return wrapper


def add(a: int, b: int):
    return a + b


print(inspect.signature(outer_function(add)))
# (a: int, b: int)
print(inspect.signature(outer_function))
# (func: Callable[..., Any])
```

**Comment by Soerboe** [comment]
@JarroVGIT I was quite happy when I found this solutions as it solves all my issues - a core feature with a dynamic plugin function. And all functions (core and plugin) can use the dependency system independently. It could also be extended to multiple levels if needed. 

I don't mind you writing a blog post on this. Please, add a link to it here so readers of this issue can find it.

**Comment by YuriiMotov** [comment, marked as answer]
For those who are interested.
You can try using this approach:
```python
import inspect
from typing import Callable, TypeVar, ParamSpec, Concatenate
from fastapi import FastAPI, Response
from functools import wraps

P = ParamSpec("P")
R = TypeVar("R", bound=str)


def add_id_header(func: Callable[P, R]) -> Callable[Concatenate[Response, P], R]:
    @wraps(func)
    def wrapper(response: Response, *args: P.args, **kwargs: P.kwargs) -> R:
        name_id = func(*args, **kwargs)
        headers = {"etag": name_id}
        response.headers.update(headers)
        return name_id

    # Modify wrapper's signature to add `response: Response`
    sig = inspect.signature(func)
    if "response" not in sig.parameters:
        new_params = list(sig.parameters.values()) + [
            inspect.Parameter(
                name="response",
                kind=inspect.Parameter.KEYWORD_ONLY,
                annotation=Response,
            )
        ]
        wrapper.__signature__ = sig.replace(parameters=new_params)  # type: ignore[attr-defined]

    return wrapper


app = FastAPI()


@app.get("/")
@add_id_header
def get_index():
    return "123"
```

Warning: this is not perfect production-ready code. Use it just to get the idea.

There is a tricky part here. Make sure you use the name for `Response` parameter consistently across all endpoint functions and other dependencies.
If you add another parameter of type `Response` but with different name it will fail to inject `Response` object
